lib/cyrusdb_twom.c

```diff
@@ -0,0 +1,2596 @@
+    header->dirty_size = dirty_size;
+    r = commit_header(db, header);
+    if (r) return r;
+
+    if (count) *count = changed;
+
+    return 0;
+}
+
+static int recovery(struct dbengine *db)
+{
+    clock_t start = sclock();
+    int count = 0;
+    int r;
+
+    /* no need to run recovery if we're consistent */
+    if (db_is_clean(db))
+        return 0;
+
+    struct tm_loc *loc = xzmalloc(sizeof(struct tm_loc));
+    loc->file = db->openfile;
+    loc->end = loc->file->header.current_size;
+    loc->file->refcount++;
+
+    r = recovery1(db, loc, &count);
+    if (r) {
+        xsyslog(LOG_ERR, "DBERROR: recovery1 failed",
+                         "filename=<%s>",
+                         db->fname);
+    }
+    else {
+        syslog(LOG_INFO,
+               "twom: recovered %s (%llu record%s, %llu bytes) in %2.3f seconds - fixed %d offset%s",
+               db->fname, (LLU)db->openfile->header.num_records,
+               db->openfile->header.num_records == 1 ? "" : "s",
+               (LLU)(db->openfile->header.current_size),
+               (sclock() - start) / (double) CLOCKS_PER_SEC,
+               count, count == 1 ? "" : "s");
+    }
+
+    loc->file->refcount--;
+    free(loc);
+
+    return r;
+}
+
+static int fetch(struct dbengine *mydb,
+                 const char *key, size_t keylen,
+                 const char **data, size_t *datalen,
+                 struct txn **tidptr)
+{
+    assert(key);
+    assert(keylen);
+    return myfetch(mydb, key, keylen, NULL, NULL,
+                   data, datalen, tidptr, 0);
+}
+
+static int fetchnext(struct dbengine *mydb,
+                 const char *key, size_t keylen,
+                 const char **foundkey, size_t *fklen,
+                 const char **data, size_t *datalen,
+                 struct txn **tidptr)
+{
+    return myfetch(mydb, key, keylen, foundkey, fklen,
+                   data, datalen, tidptr, 1);
+}
+
+static int create(struct dbengine *db,
+                  const char *key, size_t keylen,
+                  const char *data, size_t datalen,
+                  struct txn **tid)
+{
+    if (datalen) assert(data);
+    return mystore(db, key, keylen, data ? data : "", datalen, tid, 0);
+}
+
+static int store(struct dbengine *db,
+                 const char *key, size_t keylen,
+                 const char *data, size_t datalen,
+                 struct txn **tid)
+{
+    if (datalen) assert(data);
+    return mystore(db, key, keylen, data ? data : "", datalen, tid, 1);
+}
+
+static int delete(struct dbengine *db,
+                 const char *key, size_t keylen,
+                 struct txn **tid, int force)
+{
+    return mystore(db, key, keylen, NULL, 0, tid, force);
+}
+
+static int myinit(const char *dbdir __attribute__((unused)), int flags __attribute__((unused)))
+{
+    const char *checksum_engine = libcyrus_config_getstring(CYRUSOPT_TWOM_CHECKSUM_ENGINE);
+    if (!checksum_engine) return TWOM_CHECKSUM_XXH64;
```

Comment:
This looks wrong. If we *don't* have a checksum_engine, we return TWOM_CHECKSUM_XXH64, which will be treated as an error code by the caller?

Reply:
hmm, yeah.  That's totally bogus.  Luckily it got removed in the rebase (though I need to remove the associated config option now too)